### DISCUSSION TITLE
Question: Unknown entity in rule and adding facts

### DISCUSSION THREAD

**Discussion by maarek** [hyperjumptech/grule-rule-engine#180] [Q&A]
I noticed that you get quite a few errors when a rule tries to evaluate against the known facts and the current context does not have the fact or selector defined that is present in the rule. What I am trying to do is evaluate what is know and add facts to the system during execution or rerun, keeping the current context, if the current execution exhausts all paths and new facts are introduced. 

For example, if during the execution of a fact against rules, a function is called which would add new facts into the context. This would be opposed to introducing change to the currently known facts of the system in way of the known facts having some sort of map that new facts would be loaded into.

Is something like this possible or another way I should be thinking about a system like this? I should note that I'm using facts and rules from json and not struct types.

**Comment by newm4n** [comment]
Hi @maarek 

Currently, you can't add more fact into the Datacontext from within the rule. Unless, you add more fact your self from your own __helper__ function. 

For example:

```go
type Fact struct {
     Ctx *ast.DataContext
}

func (f *Fact) AddFact() {
     f.Ctx.Add("MoreFact", morefact)
}

...
...
    dataCtx := ast.NewDataContext()
    myFact := &Fact {
       Ctx : dataCtx,
    }
    dataCtx.Add("MyFact", myFact)
```

And from the GRL

```
    then
       MyFact.AddFact();  
```

**Reply by maarek** [reply]
@newm4n How do you get it to execute? Below is an example payload. I have an Initial fact that is added to the DataContext and then executed. During execution, a function call is made that will load the new fact into the same DataContext. I run into two problems with this. 

1. I get errors such as `Error while evaluating rule CheckStatusForCompletion, got left hand expression error. got left hand expression error. got non existent key NewFact`  as the NewFact has not been loaded into the system until OnInitialLoadRequest is handled.
2. `OnInitialLoadRequest` should only execute once, so I retract it. But what I never see is `CheckStatusForCompletion` being called again upon load of the new fact. Evaluation error occurred as per the above, fact is loaded, and then nothing, even if I call Execute again from the loading function.


Initial
```json
{
  "Type": "Request",
  "MyId": "1234"
}
```

Rules
```json
[
    {
        "name": "OnInitialLoadRequest",
        "desc": "on an Initial fact, load if type is request",
        "salience": 90,
        "when": {"eq": ["Initial.Type", {"const": "Request"}]},
        "then": [
            {"call": ["Retract", {"const": "OnInitialLoadRequest"}]},
            {"call": ["Log", {"const": "Loading assets by MyId"}]},
            {"call": ["Request.Request", "Initial.MyId"]},
            {"call": ["Log", {"const": "Load Completed"}]}
        ]
    },
    {
        "name": "CheckStatusForCompletion",
        "desc": "if is complete, result is complete",
        "salience": 10,
        "when":  {"lt": [{"obj": "NewFact.State"}, {"const": "Completed"}]},
        "then": [
            {"call": ["Log", {"const": "Asserting Completion"}]},
            {"set": ["Result.Facts['Status']", {"const": "Complete"}]},
            {"call": ["Retract", {"const": "CheckStatusForCompletion"}]}           
        ]
    }
]
```

NewFact
```json
{
  "State": "Completed"
}
```

**Reply by opsnull** [reply]
@newm4n same question here: How do you get it to execute ? Thanks.
